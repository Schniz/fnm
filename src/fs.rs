use std::path::Path;

#[cfg(unix)]
pub fn symlink_dir<P: AsRef<Path>, U: AsRef<Path>>(from: P, to: U) -> std::io::Result<()> {
    std::os::unix::fs::symlink(from, to)?;
    Ok(())
}

#[cfg(windows)]
pub fn symlink_dir<P: AsRef<Path>, U: AsRef<Path>>(from: P, to: U) -> std::io::Result<()> {
    junction::create(from, to)?;
    Ok(())
}

#[cfg(windows)]
pub fn remove_symlink_dir<P: AsRef<Path>>(path: P) -> std::io::Result<()> {
    std::fs::remove_dir(path)?;
    Ok(())
}

#[cfg(unix)]
pub fn remove_symlink_dir<P: AsRef<Path>>(path: P) -> std::io::Result<()> {
    std::fs::remove_file(path)?;
    Ok(())
}

pub fn shallow_read_symlink<P: AsRef<Path>>(path: P) -> std::io::Result<std::path::PathBuf> {
    std::fs::read_link(path)
}
