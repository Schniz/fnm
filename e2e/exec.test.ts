import { script } from "./shellcode/script.js"
import { Bash, Fish, PowerShell, WinCmd, Zsh } from "./shellcode/shells.js"
import testCwd from "./shellcode/test-cwd.js"
import fs from "node:fs/promises"
import path from "node:path"
import describe from "./describe.js"

for (const shell of [Bash, Zsh, Fish, PowerShell, WinCmd]) {
  describe(shell, () => {
    test("`exec` usage", async () => {
      await fs.writeFile(path.join(testCwd(), ".nvmrc"), "v8.10.0")

      await script(shell)
        .then(shell.call("fnm", ["install"]))
        .then(shell.call("fnm", ["install", "v6.10.0"]))
        .then(shell.call("fnm", ["install", "v10.10.0"]))
        .then(
          shell.hasCommandOutput(
            shell.call("fnm", ["exec", "--", "node", "-v"]),
            "v8.10.0",
            "version file exec"
          )
        )
        .then(
          shell.hasCommandOutput(
            shell.call("fnm", ["exec", "--using=6", "--", "node", "-v"]),
            "v6.10.0",
            "exec:6 node -v"
          )
        )
        .then(
          shell.hasCommandOutput(
            shell.call("fnm", ["exec", "--using=10", "--", "node", "-v"]),
            "v10.10.0",
            "exec:6 node -v"
          )
        )
        .takeSnapshot(shell)
        .execute(shell)
    })
  })
}
