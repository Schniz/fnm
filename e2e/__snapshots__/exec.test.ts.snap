// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bash \`exec\` usage: Bash 1`] = `
"set -e
fnm install
fnm install v6.10.0
fnm install v10.10.0
if [ "$(fnm exec -- node -v)" != "v8.10.0" ]; then
  echo "Expected version file exec to be v8.10.0. Got $(fnm exec -- node -v)"
  exit 1
fi
if [ "$(fnm exec --using=6 -- node -v)" != "v6.10.0" ]; then
  echo "Expected exec:6 node -v to be v6.10.0. Got $(fnm exec --using=6 -- node -v)"
  exit 1
fi
if [ "$(fnm exec --using=10 -- node -v)" != "v10.10.0" ]; then
  echo "Expected exec:6 node -v to be v10.10.0. Got $(fnm exec --using=10 -- node -v)"
  exit 1
fi"
`;

exports[`Fish \`exec\` usage: Fish 1`] = `
"fnm install
fnm install v6.10.0
fnm install v10.10.0
set ____test____ (fnm exec -- node -v)
if test "$____test____" != "v8.10.0"
  echo "Expected version file exec to be v8.10.0. Got $____test____"
  exit 1
end
set ____test____ (fnm exec --using=6 -- node -v)
if test "$____test____" != "v6.10.0"
  echo "Expected exec:6 node -v to be v6.10.0. Got $____test____"
  exit 1
end
set ____test____ (fnm exec --using=10 -- node -v)
if test "$____test____" != "v10.10.0"
  echo "Expected exec:6 node -v to be v10.10.0. Got $____test____"
  exit 1
end"
`;

exports[`PowerShell \`exec\` usage: PowerShell 1`] = `
"$ErrorActionPreference = "Stop"
fnm install
fnm install v6.10.0
fnm install v10.10.0
if ( "$(fnm exec -- node -v)" -ne "v8.10.0" ) { echo "Expected version file exec to be v8.10.0. Got $(fnm exec -- node -v)"; exit 1 }
if ( "$(fnm exec --using=6 -- node -v)" -ne "v6.10.0" ) { echo "Expected exec:6 node -v to be v6.10.0. Got $(fnm exec --using=6 -- node -v)"; exit 1 }
if ( "$(fnm exec --using=10 -- node -v)" -ne "v10.10.0" ) { echo "Expected exec:6 node -v to be v10.10.0. Got $(fnm exec --using=10 -- node -v)"; exit 1 }"
`;

exports[`Zsh \`exec\` usage: Zsh 1`] = `
"set -e
fnm install
fnm install v6.10.0
fnm install v10.10.0
if [ "$(fnm exec -- node -v)" != "v8.10.0" ]; then
  echo "Expected version file exec to be v8.10.0. Got $(fnm exec -- node -v)"
  exit 1
fi
if [ "$(fnm exec --using=6 -- node -v)" != "v6.10.0" ]; then
  echo "Expected exec:6 node -v to be v6.10.0. Got $(fnm exec --using=6 -- node -v)"
  exit 1
fi
if [ "$(fnm exec --using=10 -- node -v)" != "v10.10.0" ]; then
  echo "Expected exec:6 node -v to be v10.10.0. Got $(fnm exec --using=10 -- node -v)"
  exit 1
fi"
`;
